cpp/src/text/jaccard.cu

```diff
@@ -19,229 +19,377 @@
+  count_substrings_kernel<<<num_blocks, block_size, 0, stream.value()>>>(
+    *d_strings, width, offsets.data());
+  auto const total_hashes =
+    cudf::detail::sizes_to_offsets(offsets.begin(), offsets.end(), offsets.begin(), stream);
+
+  // hash substrings
+  rmm::device_uvector<uint32_t> hashes(total_hashes, stream);
+  substring_hash_kernel<<<num_blocks, block_size, 0, stream.value()>>>(
+    *d_strings, width, offsets.data(), hashes.data());
+
+  // sort hashes
+  rmm::device_uvector<uint32_t> sorted(total_hashes, stream);
   rmm::device_buffer d_temp_storage;
   size_t temp_storage_bytes = 0;
   cub::DeviceSegmentedSort::SortKeys(d_temp_storage.data(),
                                      temp_storage_bytes,
-                                     data,
+                                     hashes.data(),
                                      sorted.data(),
                                      sorted.size(),
                                      input.size(),
-                                     offsets,
-                                     offsets + 1,
+                                     offsets.begin(),
+                                     offsets.begin() + 1,
                                      stream.value());
   d_temp_storage = rmm::device_buffer{temp_storage_bytes, stream};
   cub::DeviceSegmentedSort::SortKeys(d_temp_storage.data(),
                                      temp_storage_bytes,
-                                     data,
+                                     hashes.data(),
                                      sorted.data(),
                                      sorted.size(),
                                      input.size(),
-                                     offsets,
-                                     offsets + 1,
+                                     offsets.begin(),
+                                     offsets.begin() + 1,
                                      stream.value());
 
-  auto contents = hashes->release();
-  // the offsets are taken from the hashes column since they are the same
-  // before and after the segmented-sort
-  return cudf::make_lists_column(
-    col.size(),
-    std::move(contents.children.front()),
-    std::make_unique<cudf::column>(std::move(sorted), rmm::device_buffer{}, 0),
-    0,
-    rmm::device_buffer{},
-    stream,
-    rmm::mr::get_current_device_resource());
+  return std::make_pair(std::move(sorted), std::move(offsets));
 }
+
+/**
+ * @brief Compute the jaccard distance for each row
+ *
+ * Formula is J = |A ∩ B| / |A ∪ B|
+ *              = |A ∩ B| / (|A| + |B| - |A ∩ B|)
+ *
+ * where |A ∩ B| is number of common values between A and B
+ * and |x| is the number of unique values in x.
+ */
+struct jaccard_fn {
+  cudf::size_type const* d_uniques1;
+  cudf::size_type const* d_uniques2;
+  cudf::size_type const* d_intersects;
+
+  __device__ float operator()(cudf::size_type idx) const
```

Comment:
It surprises me a bit that we return `float` and not `double` for this function.

Reply:
I believe this is what is expected. I can _double_ check though.

Reply:
Confirmed `float` is correct for now.